sh-utils-1.13/man/date.1
.TH DATE 1 "GNU Shell Utilities" "FSF" \" -*- nroff -*-
.SH NAME
date \- print or set the system date and time
.SH SYNOPSIS
.B date
[\-u] [\-d datestr] [\-s datestr] [\-\-utc] [\-\-universal]
[\-\-date=datestr] [\-\-set=datestr] [\-\-help] [\-\-version]
[+FORMAT] [MMDDhhmm[[CC]YY][.ss]]

.SH DESCRIPTION
This documentation is no longer being maintained and may be inaccurate
or incomplete.  The Texinfo documentation is now the authoritative source.
.PP
This manual page
documents the GNU version of
.BR date .
.B date
with no arguments prints the current time and date (in the format
of the `%c' directive described below).
If given an argument that starts with a `+', it prints the current
time and date in a format controlled by that argument, which has the
same format as the format string passed to the `strftime' function.
Except for directives that start with `%', characters in that string
are printed unchanged.
.PP
The directives are:
.IP %
a literal %
.IP n
a newline
.IP t
a horizontal tab
.PP
Time fields:
.IP %H
hour (00..23)
.IP %I
hour (01..12)
.IP %k
hour ( 0..23)
.IP %l
hour ( 1..12)
.IP %M
minute (00..59)
.IP %p
locale's AM or PM
.IP %r
time, 12-hour (hh:mm:ss [AP]M)
.IP %s
seconds since 1970-01-01 00:00:00 UTC (a nonstandard extension)
.IP %S
second (00..61)
.IP %T
time, 24-hour (hh:mm:ss)
.IP %X
locale's time representation (%H:%M:%S)
.IP %Z
time zone (e.g., EDT), or nothing if no time zone is determinable
.PP
Date fields:
.IP %a
locale's abbreviated weekday name (Sun..Sat)
.IP %A
locale's full weekday name, variable length (Sunday..Saturday)
.IP %b
locale's abbreviated month name (Jan..Dec)
.IP %B
locale's full month name, variable length (January..December)
.IP %c
locale's date and time (Sat Nov 04 12:02:33 EST 1989)
.IP %d
day of month (01..31)
.IP %D
date (mm/dd/yy)
.IP %h
same as %b
.IP %j
day of year (001..366)
.IP %m
month (01..12)
.IP %U
week number of year with Sunday as first day of week (00..53)
.IP %w
day of week (0..6) with 0 corresponding to Sunday
.IP %W
week number of year with Monday as first day of week (00..53)
.IP %x
locale's date representation (mm/dd/yy)
.IP %y
last two digits of year (00..99)
.IP %Y
year (1970...)
.PP
By default, 
.BR date
pads numeric fields with zeroes.
GNU
.BR date
recognizes the following nonstandard numeric modifiers:
.IP \-
(hyphen) do not pad the field
.IP _
(underscore) pad the field with spaces
.PP
If given an argument that does not start with `+',
.B date
sets the system clock to the time and date specified by that argument.
The argument must consist entirely of digits, which have the following
meaning:
.IP MM
month
.IP DD
day within month
.IP hh
hour
.IP mm
minute
.IP CC
first two digits of year (optional)
.IP YY
last two digits of year (optional)
.IP ss
second (optional)
.PP
Only the superuser can set the system clock.
.SS OPTIONS
.TP
.I "\-d datestr, \-\-date datestr"
Display the time and date specified in
.IR datestr ,
which can be in almost any common format.  The display is in
the default output format, or if an argument starting with `+' is
given to
.BR date ,
in the format specified by that argument.
.TP
.I "\-\-help"
Print a usage message on standard output and exit successfully.
.TP
.I "\-s datestr, \-\-set datestr"
Set the time and date to
.IR datestr ,
which can be in almost any common format.
It can contain month names, timezones, `am' and `pm', etc.
.TP
.I "\-u, \-\-universal"
Print or set the time and date in Coordinated Universal Time (also
known as Greenwich Mean Time) instead of in local (wall clock) time.
.TP
.I "\-\-version"
Print version information on standard output then exit successfully.
.SH EXAMPLES
.PP
To print the date of the day before yesterday
.IP
date \-\-date \'2 days ago\'
.PP
To print the date of the day three months and one day hence
.IP
date \-\-date \'3 months 1 day\'
.PP
To print the day of year of Christmas in the current year
.IP
date --date \'25 Dec\' +%j
.PP
To print the current date in a format including the full month name and
the day of the month
.IP
date \'+%B %d\'
.PP
But this may not be what you want because for the first nine days
of the month, the \`%d\' expands to a zero-padded two-digit field,
for example \`date \-d 1-may \'+%B %d\'\' will print
\`May 01\'.
.PP
To print the same date but without the leading zero for one-digit
days of month, you can use the nonstandard \`\-\' modifier to suppress
the padding altogether.
.IP
date \-d 1-may \'+%B %-d\'


m4-1.4/NEWS
GNU m4 NEWS - User visible changes.
Copyright (C) 1992, 1993, 1994 Free Software Foundation, Inc.

Version 1.4 - October 1994, by Franc,ois Pinard

(No user visible changes)

Version 1.3 - September 1994, by Franc,ois Pinard

* Diversions are created as needed.  Option `-N' is still accepted, but
otherwise ignored.  Users should use only negative diversion numbers,
instead of high positive numbers, for diverting to nowhere.

* Diversions should also work faster.  No temporary files will be needed
at all if all diversions taken altogether do not use more than 512K.

* Frozen state files may be produced with the `--freeze-state' (-F)
option and later brought back through the `--reload-state' (-R) option.

Version 1.2 - July 1994, by Franc,ois Pinard

* In patsubst(STRING, REGEXP, REPLACEMENT), \& in REPLACEMENT has been
changed to represent this part of STRING matched by the whole REGEXP,
instead of the whole STRING as before.  \0 does the same, but emits a
diagnostic saying it will disappear in some subsequent release.

* eval(EXPR) emits a diagnostic if EXPR has suffixed crumb.  The same for
other numeric conversions in incr(), decr(), divert(), etc.

* `--fatal-warnings' (-E) stops execution at first warning.

* `--nesting-limit=LEVEL' (-L LEVEL) sets a limit to macro nesting.
It is initially fixed at 250.

* `--word-regexp=REGEXP' (-W REGEXP) modifies macro name syntax, like
does the new `changeword(REGEXP)' macro.  This feature is experimental,
tell me your opinions about it.  You do need --enable-changeword at
configure time to get these things.  Do *not* depend on them yet.

* Trace output format is scannable by GNU Emacs' next-error function.

* Stack overflow is detected and diagnosed on some capable systems.

* Various bugs have been corrected, m4 should be more portable.  See the
ChangeLog for details.

Version 1.1 - November 1993, by Franc,ois Pinard

Changes which might affect existing GNU m4 scripts:

* Option `-V' has been removed, use `--version' instead.  `--version'
writes on standard output instead of standard error, and inhibits any
script execution.

* `--no-gnu-extensions' has been renamed `--traditional'.

* In `eval', `^' used to indicate exponentiation, use `**' instead.

* The automatic undiversion which takes place at end of all input is
forced into the main output stream.

Changes which are unlikely to affect existing scripts:

* `--help' prints an usage summary on standard output.  Script execution
is then inhibited.

* `--prefix-builtins' (-P) prefixes all builtin macros by `m4_'.

* Most builtin macros for which arguments are mandatory, called without
any arguments, are no more recognized as builtin macros: they are
consequently copied verbatim to the output stream.

* `define' and `pushdef' are usable with only one argument, they give
this argument an empty definition.

* `eval' new operators for binary representation handling: `^' for
exclusive-or, `~' for the bitwise negation, `<<' and `>>' for shifts.

* `eval' recognizes the notation 0bDIGITS for binary numbers and the
notation 0rRADIX:DIGITS for numbers in any radix from 1 to 36.

Version 1.0.3 - December 1992, by Franc,ois Pinard

Changes for the user:

* `dnl' outputs a diagnostic if immediately followed by `('.  Usually,
`dnl' is followed by newline or whitespace.

* `ifelse' accepts without complaining the common idiom of having only
one argument.  This is useful for introducing long comments.

* `eval' always expresses values as signed, whatever the radix.

* M4OPTS environment variable is no longer obeyed.

* `--no-warnings' option is renamed `--silent'.

* Debug lines use a new format more compatible with GNU standards.

* Various bugs have been corrected.  See the ChangeLog for details.

Changes for the installer:

* GNU m4 now uses an Autoconf-generated configure script, and should be
more easily portable in many ways.  (Cray is not supported yet).

* `make check' has been made more portable, expect no errors.

Changes for the programmer:

* Sources have been fully reindented to comply with GNU standards, and
cleaned up in many ways.

* Sources have been protoized.  Non-ANSI compilers are automatically
detected, then sources are unprotoized on the fly before compilation.

* GNU m4 uses newer versions of obstack, regex, getopt, etc.

Version 1.0 - October 1991, by Rene' Seindal

* Uses GNU configure, taken from the gdb distribution.

* Uses GNU getopt(), with long option names.

* The -Q/+quiet option is added, which suppresses warnings about missing
or superflous arguments to built-in macros.

* Added default options via the M4OPTS environment variable.

* Several minor bugs have been fixed.

Version 0.99 - July 1991, by Rene' Seindal

* The builtins `incr' and `decr' are now implemented without use of
`eval'.

* The builtin `indir' is added, to allow for indirect macro calls
(allows use of "illegal" macro names).

* The debugging and tracing facilities has been enhanced considerably.
See the manual for details.

* The -tMACRO option is added, marks MACRO for tracing as soon as it
is defined.

* Builtins are traced after renaming iff they were before.

* Named files can now be undiverted.

* The -Nnum option can be used to increase the number of divertions
available.

* Calling changecom without arguments now disables all comment handling.

* A bug in `dnl' is fixed.

* A bug in the multi-character quoting code is fixed.

* Several typos in the manual has been corrected.  More probably persist.

Version 0.75 - November 1990, by Rene' Seindal

* Implemented search path for include files (-I option and M4PATH
environment variable).

* Implemented builtin `format' for printf-like formatting.

* Implemented builtin `regexp' for searching for regular expressions.


shellutils-1.9.4/src/date.c
/* date - print or set the system date and time
   Copyright (C) 89, 90, 91, 92, 93, 1994 Free Software Foundation, Inc.

   This program is free software; you can redistribute it and/or modify
   it under the terms of the GNU General Public License as published by
   the Free Software Foundation; either version 2, or (at your option)
   any later version.

   This program is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.

   You should have received a copy of the GNU General Public License
   along with this program; if not, write to the Free Software
   Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.  */

/* Options:
   -d DATESTR	Display the date DATESTR.
   -s DATESTR	Set the date to DATESTR.
   -u		Display or set the date in universal instead of local time.
   +FORMAT	Specify custom date output format, described below.
   MMDDhhmm[[CC]YY][.ss]	Set the date in the format described below.

   If one non-option argument is given, it is used as the date to which
   to set the system clock, and must have the format:
   MM	month (01..12)
   DD	day in month (01..31)
   hh	hour (00..23)
   mm	minute (00..59)
   CC	first 2 digits of year (optional, defaults to current) (00..99)
   YY	last 2 digits of year (optional, defaults to current) (00..99)
   ss	second (00..61)

   If a non-option argument that starts with a `+' is specified, it
   is used to control the format in which the date is printed; it
   can contain any of the `%' substitutions allowed by the strftime
   function.  A newline is always added at the end of the output.

   David MacKenzie <djm@gnu.ai.mit.edu> */

#ifdef HAVE_CONFIG_H
#if defined (CONFIG_BROKETS)
/* We use <config.h> instead of "config.h" so that a compilation
   using -I. -I$srcdir will use ./config.h rather than $srcdir/config.h
   (which it would do because it found this file in $srcdir).  */
#include <config.h>
#else
#include "config.h"
#endif
#endif

#include <stdio.h>
#include <getopt.h>
#include <sys/types.h>

#include "version.h"
#include "system.h"

#ifdef TM_IN_SYS_TIME
#include <sys/time.h>
#else
#include <time.h>
#endif

#ifndef STDC_HEADERS
size_t strftime ();
time_t time ();
#endif

int putenv ();
int stime ();

char *xrealloc ();
time_t get_date ();
time_t posixtime ();
void error ();

static void show_date ();
static void usage ();

/* putenv string to use Universal Coordinated Time.
   POSIX.2 says it should be "TZ=UCT0" or "TZ=GMT0". */
#ifndef TZ_UCT
#if defined(hpux) || defined(__hpux__) || defined(ultrix) || defined(__ultrix__) || defined(USG)
#define TZ_UCT "TZ=GMT0"
#else
#define TZ_UCT "TZ="
#endif
#endif

/* The name this program was run with, for error messages. */
char *program_name;

/* If non-zero, display usage information and exit.  */
static int show_help;

/* If non-zero, print the version on standard output and exit.  */
static int show_version;

static struct option const long_options[] =
{
  {"date", required_argument, NULL, 'd'},
  {"help", no_argument, &show_help, 1},
  {"set", required_argument, NULL, 's'},
  {"uct", no_argument, NULL, 'u'},
  {"universal", no_argument, NULL, 'u'},
  {"version", no_argument, &show_version, 1},
  {NULL, 0, NULL, 0}
};

void
main (argc, argv)
     int argc;
     char **argv;
{
  int optc;
  char *datestr = NULL;
  time_t when;
  int set_date = 0;
  int universal_time = 0;

  program_name = argv[0];

  while ((optc = getopt_long (argc, argv, "d:s:u", long_options, (int *) 0))
	 != EOF)
    switch (optc)
      {
      case 0:
	break;
      case 'd':
	datestr = optarg;
	break;
      case 's':
	datestr = optarg;
	set_date = 1;
	break;
      case 'u':
	universal_time = 1;
	break;
      default:
	usage (1);
      }

  if (show_version)
    {
      printf ("%s\n", version_string);
      exit (0);
    }

  if (show_help)
    usage (0);

  if (argc - optind > 1)
    usage (1);

  if (universal_time && putenv (TZ_UCT) != 0)
    error (1, 0, "virtual memory exhausted");

  time (&when);

  if (datestr)
    when = get_date (datestr, NULL);

  if (argc - optind == 1 && argv[optind][0] != '+')
    {
      when = posixtime (argv[optind]);
      set_date = 1;
    }

  if (when == -1)
    error (1, 0, "invalid date");

  if (set_date && stime (&when) == -1)
    error (0, errno, "cannot set date");

  if (argc - optind == 1 && argv[optind][0] == '+')
    show_date (argv[optind] + 1, when);
  else
    show_date ((char *) NULL, when);

  exit (0);
}

/* Display the date and/or time in WHEN according to the format specified
   in FORMAT, followed by a newline.  If FORMAT is NULL, use the
   standard output format (ctime style but with a timezone inserted). */

static void
show_date (format, when)
     char *format;
     time_t when;
{
  struct tm *tm;
  char *out = NULL;
  size_t out_length = 0;

  tm = localtime (&when);

  if (format == NULL)
    /* Print the date in the default format.  Vanilla ANSI C strftime
       doesn't support %e, but POSIX requires it.  If you don't use
       a GNU strftime, make sure yours supports %e.  */
    format = "%a %b %e %H:%M:%S %Z %Y";
  else if (*format == '\0')
    {
      printf ("\n");
      return;
    }

  do
    {
      out_length += 200;
      out = (char *) xrealloc (out, out_length);
    }
  while (strftime (out, out_length, format, tm) == 0);

  printf ("%s\n", out);
  free (out);
}

static void
usage (status)
     int status;
{
  if (status != 0)
    fprintf (stderr, "Try `%s --help' for more information.\n",
	     program_name);
  else
    {
      printf ("Usage: %s [OPTION]... [+FORMAT] [MMDDhhmm[[CC]YY][.ss]]\n",
	      program_name);
      printf ("\
\n\
  -d, --date=STRING        display time described by STRING, not `now'\n\
  -s, --set=STRING         set time described by STRING\n\
  -u, --uct, --universal   print or set Universal Coordinated Time\n\
      --help               display this help and exit\n\
      --version            output version information and exit\n\
");
      printf ("\
\n\
FORMAT controls the output.  Interpreted sequences are:\n\
\n\
  %%%%   a literal %%\n\
  %%a   locale's abbreviated weekday name (Sun..Sat)\n\
  %%A   locale's full weekday name, variable length (Sunday..Saturday)\n\
  %%b   locale's abbreviated month name (Jan..Dec)\n\
  %%B   locale's full month name, variable length (January..December)\n\
  %%c   locale's date and time (Sat Nov 04 12:02:33 EST 1989)\n\
  %%d   day of month (01..31)\n\
  %%D   date (mm/dd/yy)\n\
  %%h   same as %%b\n\
  %%H   hour (00..23)\n\
  %%I   hour (01..12)\n\
  %%j   day of year (001..366)\n\
  %%k   hour ( 0..23)\n\
  %%l   hour ( 1..12)\n\
  %%m   month (01..12)\n\
  %%M   minute (00..59)\n\
  %%n   a newline\n\
  %%p   locale's AM or PM\n\
  %%r   time, 12-hour (hh:mm:ss [AP]M)\n\
  %%s   seconds since 00:00:00, Jan 1, 1970 (a nonstandard extension)\n\
  %%S   second (00..61)\n\
  %%t   a horizontal tab\n\
  %%T   time, 24-hour (hh:mm:ss)\n\
  %%U   week number of year with Sunday as first day of week (00..53)\n\
  %%w   day of week (0..6);  0 represents Sunday\n\
  %%W   week number of year with Monday as first day of week (00..53)\n\
  %%x   locale's date representation (mm/dd/yy)\n\
  %%X   locale's time representation (%%H:%%M:%%S)\n\
  %%y   last two digits of year (00..99)\n\
  %%Y   year (1970...)\n\
  %%Z   time zone (e.g., EDT), or nothing if no time zone is determinable\n\
");
    }
  exit (status);
}
